src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -479,13 +479,14 @@ private BoundExpression MissingDeconstruct(BoundExpression receiver, ExpressionS
             return BadExpression(syntax, childNode);
         }
 
-        private BoundLocalDeconstructionDeclaration BindDeconstructionDeclarationStatementParts(LocalDeclarationStatementSyntax node, DiagnosticBag diagnostics)
+        internal BoundLocalDeconstructionDeclaration BindDeconstructionDeclaration(StatementSyntax statement, VariableDeclarationSyntax declaration, DiagnosticBag diagnostics)
         {
-            Debug.Assert(node.Declaration.Deconstruction != null);
+            Debug.Assert(statement.Kind() == SyntaxKind.LocalDeclarationStatement || statement.Kind() == SyntaxKind.ForStatement);
+            Debug.Assert(declaration.Deconstruction != null);
 
-            ArrayBuilder<DeconstructionVariable> variables = BindDeconstructionDeclarationVariables(node.Declaration, node.Declaration.Type, diagnostics);
+            ArrayBuilder<DeconstructionVariable> variables = BindDeconstructionDeclarationVariables(declaration, declaration.Type, diagnostics);
 
-            var result = new BoundLocalDeconstructionDeclaration(node, BindDeconstructionAssignment(node.Declaration.Deconstruction.Value, node.Declaration.Deconstruction.Value, variables, diagnostics));
+            var result = new BoundLocalDeconstructionDeclaration(statement, BindDeconstructionAssignment(declaration.Deconstruction.Value, declaration.Deconstruction.Value, variables, diagnostics));
```

Comment:
There is a check in `BindStatement` (in `Binder_Statement.cs` line 132) that all bound statements should be associated with a statement syntax.
In the for statement, the `VariableDeclaration` used to initialize the loop is not a statement (it's just an expression).